### PR TITLE
kernel: Add k_heap_calloc

### LINKED ISSUE #79417: kernel: Missing calloc equivalent funciton
**Is your enhancement proposal related to a problem? Please describe.**
The Zephyr kernel contains various allocation functions: `k_heap_alloc`, `k_heap_realloc`, `k_heap_free`, etc.
Those can be seen as wrappers to the standard C functions `malloc`, `realloc`, `free`, etc.
However, there is no equivalent function to `calloc`.

**Describe the solution you'd like**
An additional allocation function could be added alongside the already present ones.

### PR DESCRIPTION
**Pull request by sorru94**
Add the `k_heap_calloc` function. Equivalent to the standard C `calloc` function.

Closes #79417.

### CODE REVIEW

File: subsys/tracing/sysview/SYSVIEW_Zephyr.txt [hunk 1]
```diff
@@ -82,6 +82,7 @@ TaskState 0xBF 1=dummy, 2=Waiting, 4=New, 8=Terminated, 16=Suspended, 32=Termina
 
 76  k_heap_init                heap=%I, mem=%p, bytes=%u
 77  k_heap_alloc               heap=%I, bytes=%u, Timeout=%TimeOut | Returns %p
+77  k_heap_calloc              heap=%I, num=%u, size=%u, Timeout=%TimeOut | Returns %p
```
Comment (cfriedt) [inline subsys/tracing/sysview/SYSVIEW_Zephyr.txt:85]: Can this change be added in a separate commit? It's not exactly part of the implementation

Comment (cfriedt) [reply subsys/tracing/sysview/SYSVIEW_Zephyr.txt:85]: Thanks!

File: kernel/kheap.c [hunk 2]
```diff
@@ -116,6 +116,25 @@ void *k_heap_alloc(struct k_heap *heap, size_t bytes, k_timeout_t timeout)
...
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_heap, calloc, heap, timeout);
+
+	void *ret = NULL;
+	size_t bounds = 0U;
+
+	if (!size_mul_overflow(num, size, &bounds)) {
+		ret = k_heap_alloc(heap, bounds, timeout);
+	}
+	if (ret != NULL) {
+		(void)memset(ret, 0, bounds);
```
Comment (dcpleung) [inline kernel/kheap.c:130]: This may not work correctly if the allocation alignment is bigger than `(num * size)`. For example, you need 2 bytes but the alignment is 8 bytes (pointer size for 64-bit system), this memset will not clear the last 6 bytes.

Comment (sorru94) [reply kernel/kheap.c]: I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to  `sizeof(void *)`.
I would use `sizeof(void *)` for memsetting. 
What do you think about it?

**Comment by npitre** [comment]
On Tue, 8 Oct 2024, Simone Orru wrote:

> I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to  `sizeof(void *)`.
> I would use `sizeof(void *)` for memsetting. 
> What do you think about it?

I think this is wrong. In fact I'm not sure I understand the initial 
comment either.

You should zero the requested memory not more. The user is not expected 
toconsume more memory than it asked. The fact that the allocator may 
round up the allocation is an internal detail that no one should rely 
upon.

**Comment by sorru94** [comment]
> On Tue, 8 Oct 2024, Simone Orru wrote: I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to `sizeof(void *)`. I would use `sizeof(void *)` for memsetting. What do you think about it?
> I think this is wrong. In fact I'm not sure I understand the initial comment either. You should zero the requested memory not more. The user is not expected toconsume more memory than it asked. The fact that the allocator may round up the allocation is an internal detail that no one should rely upon.

I can revert the change. But let's first make sure which one of the two is the correct implementation.

**Comment by npitre** [comment]
On Tue, 8 Oct 2024, Daniel Leung wrote:

> Part of me thinks not zeroing out the whole block may have security implications. Maybe it's not an issue.

Security critical code should normally zero its memory before freeing 
it. You can't count on the next memory requestor to do it for you. And 
in that case, too, you're not supposed to access beyond what you 
initially asked for.

calloc() is just a convenience for getting already zeroed memory.

**Review by cfriedt — CHANGES_REQUESTED** [review]
There seems to be quite a bit of unnecessary code-formatting changes. Please only change code that is relevant to the PR.

**Comment by npitre** [comment]
> E.g. the case that Daniel pointed out regarding `memset()`.

As mentioned, that particular case is wrong.

**Comment by sorru94** [comment]
> There seems to be quite a bit of unnecessary code-formatting changes. Please only change code that is relevant to the PR.

The format changes in `subsys/tracing/sysview/tracing_sysview_ids.h` are all related to the alignment of defines. 
Since my changes are only to some of those defines if I only perform alignment for my changes it will result in a header which is partially aligned.

**Comment by cfriedt** [comment]
Sure - as long as it's a separate commit, that's ok.

**Review by cfriedt — DISMISSED** [review]
LGTM

File: tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c [hunk 3]
```diff
@@ -265,3 +267,27 @@ ZTEST(k_heap_api, test_k_heap_alloc_pending_null)
 
 	k_heap_free(&k_heap_test, p);
 }
+
+/**
+ * @brief Test to demonstrate k_heap_calloc() and k_heap_free() API usage
+ *
+ * @ingroup kernel_kheap_api_tests
+ *
+ * @details The test allocates 256 unsigned integers of 32 bytes for a
```
Comment (peter-mitsis) [inline tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c:276]: Nit: I think you mean 32 bits (or 4 bytes)

**Review by peter-mitsis — DISMISSED** [review]
One minor nit, but I don't think that is a stopper.

**Review by TaiJuWu — DISMISSED** [review]

**Comment by carlescufi** [comment]
@sorru94 can you please rebase?

**Review by TaiJuWu — DISMISSED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by kartben** [comment]
@andyross @peter-mitsis please help review

**Review by npitre — APPROVED** [review]

**Review by andyross — APPROVED** [review]
